src/Razor/src/Microsoft.AspNetCore.Razor.Common/Telemetry/TelemetryReporter.cs

```diff
@@ -46,8 +47,52 @@ public void ReportEvent<T>(string name, TelemetrySeverity severity, ImmutableDic
         Report(telemetryEvent);
     }
 
-    private static string GetTelemetryName(string name) => "razor/" + name;
-    private static string GetPropertyName(string name) => "razor." + name;
+    public void ReportFault(Exception exception, string? message, object[] @params)
+    {
+        try
+        {
+            if (exception is OperationCanceledException { InnerException: { } oceInnerException })
+            {
+                ReportFault(oceInnerException, message, @params);
+                return;
+            }
+
+            if (exception is AggregateException aggregateException)
+            {
+                // We (potentially) have multiple exceptions; let's just report each of them
+                foreach (var innerException in aggregateException.Flatten().InnerExceptions)
+                {
+                    ReportFault(innerException, message, @params);
+                }
+
+                return;
+            }
+
+            var currentProcess = Process.GetCurrentProcess();
+
+            var faultEvent = new FaultEvent(
+                eventName: GetTelemetryName("fault"),
+                description: GetDescription(exception),
+                FaultSeverity.General,
+                exceptionObject: exception,
+                gatherEventDetails: faultUtility =>
+                {
+                    // Returning "0" signals that, if sampled, we should send data to Watson.
+                    // Any other value will cancel the Watson report. We never want to trigger a process dump manually,
+                    // we'll let TargetedNotifications determine if a dump should be collected.
+                    // See https://aka.ms/roslynnfwdocs for more details
+                    return 0;
+                });
+
+            Report(faultEvent);
+        }
+        catch (Exception)
+        {
+        }
+    }
+
+    private static string GetTelemetryName(string name) => "dotnet/razor/" + name;
+    private static string GetPropertyName(string name) => "dotnet.razor." + name;
```

Comment:
@ryzngard note our telemetry events and property names have changed, in case you need to update any dashboards